File: webauthn-server-core/src/main/scala/com/yubico/webauthn/data/AuthenticationDataFlags.scala
package com.yubico.webauthn.data

import com.fasterxml.jackson.annotation.JsonProperty


case class AuthenticationDataFlags(value: Byte) {
  /** User present */
  @JsonProperty
  def UP: Boolean = (value & 0x01) > 0

  /** User verified */
  @JsonProperty
  def UV: Boolean = (value & 0x04) > 0

  /** Attestation data present */
  @JsonProperty
  def AT: Boolean = (value & 0x40) > 0

  /** Extension data present */
  @JsonProperty
  def ED: Boolean = (value & 0x80) > 0

  /* Reserved bits */
  // def RFU1: Boolean = (value & 0x02) > 0
  // def RFU2_1: Boolean = (value & 0x08) > 0
  // def RFU2_2: Boolean = (value & 0x10) > 0
  // def RFU2_3: Boolean = (value & 0x20) > 0
}

File: webauthn-server-core/src/main/java/com/yubico/webauthn/data/AuthenticationDataFlags.java
package com.yubico.webauthn.data;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import lombok.EqualsAndHashCode;
import lombok.ToString;

@ToString
@EqualsAndHashCode
public final class AuthenticationDataFlags {
    public final byte value;

    /** User present */
    public final boolean UP;

    /** User verified */
    public final boolean UV;

    /** Attestation data present */
    public final boolean AT;

    /** Extension data present */
    public final boolean ED;

    @JsonCreator
    public AuthenticationDataFlags(@JsonProperty("value") byte value) {
        this.value = value;

        UP = (value & 0x01) > 0;
        UV = (value & 0x04) > 0;
        AT = (value & 0x40) > 0;
        ED = (value & 0x80) > 0;
    }

    /* Reserved bits */
    // public final boolean RFU1 = (value & 0x02) > 0;
    // public final boolean RFU2_1 = (value & 0x08) > 0;
    // public final boolean RFU2_2 = (value & 0x10) > 0;
    // public final boolean RFU2_3 = (value & 0x20) > 0;
}
